Source/WebCore/html/HTMLElement.cpp

```diff
@@ -1578,6 +1578,49 @@ void HTMLElement::popoverAttributeChanged(const AtomString& value)
         ensurePopoverData().setPopoverState(newPopoverState);
 }
 
+static const AtomString& togglePopoverAtom()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("togglepopover"_s);
+    return identifier;
+}
+
+static const AtomString& showPopoverAtom()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("showpopover"_s);
+    return identifier;
+}
+
+static const AtomString& hidePopoverAtom()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("hidepopover"_s);
+    return identifier;
+}
+
+bool HTMLElement::handleInvokeInternal(const HTMLFormControlElement* invoker, const AtomString& action)
+{
+    if (popoverState() == PopoverState::None)
+        return false;
+
+    bool shouldShowPopover = equalIgnoringASCIICase(action, autoAtom())
+        || equalIgnoringASCIICase(action, togglePopoverAtom())
+        || equalIgnoringASCIICase(action, showPopoverAtom());
+    bool shouldHidePopover = equalIgnoringASCIICase(action, autoAtom())
+        || equalIgnoringASCIICase(action, togglePopoverAtom())
+        || equalIgnoringASCIICase(action, hidePopoverAtom());
```

Comment:
Calling `equalIgnoringASCIICase` does not require or benefit from `AtomString`, so the global storage above is wasted. It would be better to have `ASCIILiteral` constants, which would not require `MainThreadNeverDestroyed` and so be faster and use less storage, and use `equalLettersIgnoringASCIICase`, which is faster. It’s OK to use `AtomString` when we want to do a case sensitive comparison, because that gets optimized to a pointer compare.

We should only evaluate one of these booleans, rather than always computing both but only using one. Something like this:
```
if (isPopoverShowing()) {
    bool shouldHide = equalLettersIgnoringASCIICase …
    if (shouldHide) {
        hidePopover();
        return true;
    }
} else {
    bool shouldShow = equalLettersIgnoringASCIICase …
    if (shouldShow) {
        showPopover(invoker);
        return true;
    }
}
```